students/lauraannf/Lesson05/mailroom_3.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Oct 30 11:18:22 2018
+
+@author: Laura.Fiorentino
+"""
+import datetime
+import sys
+from pathlib import Path
+my_path = 'C:\\Users\\Laura.Fiorentino\\Documents\\MyPython\\\
+Self_Paced-Online\\students\\lauraannf\\Lesson05\\'
+donor_list = {
+    'Frank Reynolds': [10, 20, 50],
+    'Dee Reynolds': [25, 100],
+    'Dennis Reynolds': [10, 50],
+    'Mac McDonald': [25, 35, 20],
+    'Charlie Kelly': [0.25]
+    }
+
+
+def add_donation(name, donation):
+    donor_list[name].append(donation)
+
+
+def create_email():
+    while True:
+        name = input("Type full donor name, ''list'' for choices, or \
+'quit' to quit>")
+        if name.lower() == 'list':
+            print(donor_list.keys())
+        elif name.lower() == 'quit':
+            quit_program()
+        else:
+            donation = float(input("Donation Amount?>"))
+            try:
+                add_donation(name, donation)
+            except KeyError:
+                donor_list[name] = []
+                add_donation(name, donation)
+            finally:
+                print('writing the thank you letter...')
+                write_email(name, donation)
+            break
+
+
+def create_report():
+    print('-------List of Donors-------')
+    print('{:<20}{:<20}{:<20}{:<20}'.format('Donor Name', 'Total Donated',
+                                            '# of donations',
+                                            'Average donation'))
+    print('-----------------   '*4)
+    for key in donor_list.keys():
+        print('{:<20}${:<20.2f}{:<20d}${:<20.2f}'.format(key,
+                                                         sum(donor_list[key]),
+                                                         len(donor_list[key]),
+                                                         sum(donor_list[key])
+                                                         / len(donor_list[key]
+                                                               )))
```

Comment:
@charlesrobison in this particular case it is better to leave as a for loop for couple reasons:
* comprehensions are for sequence generation, in this case print statement does not return anything so you would accumulate a sequence of None values
* this would be quite long comprehension statement and lose readability


some mailrooms don't have a potential to use comprehensions and it's ok 🙂